src/components/SearchAndFilter/SearchAndFilter.test.tsx

```diff
@@ -145,125 +167,201 @@ describe("Search and filter", () => {
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
+        "Unknown value"
+      );
+    });
+    // eslint-disable-next-line testing-library/no-node-access
+    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
+      0
+    );
   });
 
-  it("correct number of panels appear for matching search terms", () => {
+  it("correct number of panels appear for matching search terms", async () => {
     const returnSearchData = jest.fn();
-    const wrapper = mount(
+    render(
       <SearchAndFilter
         filterPanelData={sampleData}
         returnSearchData={returnSearchData}
       />
     );
-    wrapper.find(".p-search-and-filter").simulate("click");
-    wrapper
-      .find(".p-search-and-filter__input")
-      .simulate("change", { target: { value: "Google" } });
-    expect(wrapper.find(".p-filter-panel-section").length).toEqual(1);
-    wrapper
-      .find(".p-search-and-filter__input")
-      .simulate("change", { target: { value: "re" } });
-    expect(wrapper.find(".p-filter-panel-section").length).toEqual(2);
+    await waitFor(async () => {
+      await userEvent.click(
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
+      );
+    });
+    await waitFor(async () => {
+      await userEvent.clear(
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
+      );
+    });
+    await waitFor(async () => {
+      await userEvent.type(
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
+        "Google"
+      );
+    });
+    // eslint-disable-next-line testing-library/no-node-access
+    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
+      1
+    );
+    await waitFor(async () => {
+      await userEvent.clear(
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
+      );
+    });
+    await waitFor(async () => {
+      await userEvent.type(
+        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
+        "re"
+      );
+    });
+    // eslint-disable-next-line testing-library/no-node-access
+    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
+      2
+    );
   });
 
-  it("Matching search terms are highlighted with strong tag", () => {
+  it("Matching search terms are highlighted with strong tag", async () => {
     const returnSearchData = jest.fn();
-    const wrapper = mount(
+    render(
       <SearchAndFilter
         filterPanelData={sampleData}
         returnSearchData={returnSearchData}
       />
     );
-    wrapper.find(".p-search-and-filter__input").simulate("click");
-    wrapper
-      .find(".p-search-and-filter__input")
-      .simulate("change", { target: { value: "Google" } });
-    const firstChip = wrapper.find(".p-chip").first();
-    expect(firstChip.find(".p-chip__value").html()).toEqual(
-      '<span class="p-chip__value"><strong>Google</strong></span>'
-    );
+    await waitFor(async () => {
```

Comment:
Many of these tests are getting quite verbose because of these events.

Is it neccessary to wrap all of them individually in `await waitFor`? Even if the content itself is `await userEvent.something…` ?

And also, we tent to query for the same element over and over again (like `screen.getByRole("searchbox", { name: Label.SearchAndFilter })`). Could we do it once per test and store in in a variable?

Reply:
> Many of these tests are getting quite verbose because of these events.
> 
> Is it neccessary to wrap all of them individually in `await waitFor`? Even if the content itself is `await userEvent.something…` ?

As I understand it there are two different events these are waiting for. The `await userEvent.something…` is waiting for the click event but doesn't wait for other effects to complete. In this component once the panel is shown there is also a `useEffect` that responds to the state update. This state update then needs to be waited for hence the `await waitFor`.

I just had another try using things like `findBy` which is async, but I couldn't find another way to prevent the async errors.

> 
> And also, we tent to query for the same element over and over again (like `screen.getByRole("searchbox", { name: Label.SearchAndFilter })`). Could we do it once per test and store in in a variable?

I've moved all these instance to a function to get the element.